### PR TITLE
158 persist irv interpretations for acvr uploads

### LINKED ISSUE #158: Persist IRV interpretations for ACVR uploads.
Depends on #84 and #100.
Persistence is already done for the initially-uploaded cvrs. Add persistence for auditor-uploaded IRV ballots, when they're not valid IRV votes.

### PR DESCRIPTION
**Pull request by vteague**
This PR includes some useful testing of the ACVR upload endpoint, which might be useful for other tests.

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/ACVRUpload.java [hunk 1]
```diff
@@ -38,6 +43,9 @@
  *
  * @author Daniel M. Zimmerman <dmz@freeandfair.us>
  * @version 1.0.0
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/ACVRUpload.java:45]: IRVBallotIntepretation table (minus the s)

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/ACVRUpload.java [hunk 2]
```diff
@@ -200,4 +217,38 @@ public String endpointBody(final Request the_request, final Response the_respons
...
+   * The actual interpretation of the audit cvr, for audit computations, is made in
+   * CVRContestInfoJsonAdapter - this function is just for record keeping.
+   * @param cvr                  the uploaded audit cvr.
+   * @param recordType           should be either AUDITOR_ENTERED or REAUDIT.
+   * @param cvrNumber            the cvr number (as printed on the csv output).
+   * @param imprintedID          the imprinted ID.
+   * @throws IRVParsingException if the choices cannot be interpreted as IRV choices, i.e. of the
+   *                             form name(r) for integer rank r. Note this is separate from whether
+   *                             they are a valid list of IRV choices.
+   */
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/ACVRUpload.java:234]: I think you can make cvrNumber final as well.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/CVRContestInfo.java [hunk 3]
```diff
@@ -150,7 +173,12 @@ public ConsensusValue consensus() {
   public List<String> choices() {
     return Collections.unmodifiableList(my_choices);
   }
-  
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/model/CVRContestInfo.java:175]: Note in the method comment that it returns an unmodifiable list.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+   * Init mocks, particularly for authentication.
+   */
+  @BeforeClass
+  public void initMocks() {
+    testUtils.log(LOGGER, "initMocks");
+
+    // Mock successful auth as a county.
+    mockAuth(countyName, countyID);
+  }
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java:331]: Some of the list items have commas at the end and some don't, which is a bit inconsistent.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+      uploadEndpoint.before(request, response);
+
+      // Before the test, there should be 10 UPLOADED and zero AUDITOR_ENTERED cvrs.
+      final List<CastVoteRecord> preCvrs
+          = CastVoteRecordQueries.getMatching(1L, CastVoteRecord.RecordType.UPLOADED).toList();
+      final List<CastVoteRecord> preACvrs
+          = CastVoteRecordQueries.getMatching(1L, AUDITOR_ENTERED).toList();
+      assertEquals(preCvrs.size(), 10);
+      assertEquals(preACvrs.size(), 0);
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java:366]: There is something odd with the comments in the following that start with two // //

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+    }
+  }
+
+  /**
+   * Test that there is exactly one record matching the given CvrID, imprinted ID, and contest name. Check that it has
+   * the expected raw choices and valid interpretation.
+   * @param CvrNum              The cvr record number (same as the last number in the imprinted ID).
+   * @param imprintedId         The imprinted ID (scanner-batch-record).
+   * @param rawChoices          The raw choices (with parentheses, presumed invalid).
+   * @param validInterpretation The valid interpretation of the raw choices.
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/ACVRUploadTests.java:409]: Primitive types can be declared final in the input arguments.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Set;
+
+/**
+ * A class that can behave like a Spark request, for testing endpoints.
+ * Note this does _not_ (yet) implement all the functions you might need - it contains only the ones
+ * that I noticed being used in colorado-rla.
+ * https://javadoc.io/doc/com.sparkjava/spark-core/2.5.4/spark/Request.html
+ */
+public class SparkRequestStub extends Request {
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java:35]: Can we add some further commenting to the attributes/methods in this class?

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkResponseStub.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import spark.Response;
+
+/**
+ * A class that can behave like a Spark request, for testing endpoints.
+ * Note this does _not_ (yet) implement all the functions you might need - it contains only the ones
+ * that I noticed being used in colorado-rla.
+ * There are others, including redirect(), type() and raw() that are not yet implemented.
+ * See https://javadoc.io/doc/com.sparkjava/spark-core/2.5.4/spark/Response.html for types.
+ */
+public class SparkResponseStub extends Response {
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkResponseStub.java:33]: Can we have more comments for attributes/methods?

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/TestClassWithAuth.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  protected AuthenticationInterface auth;
+
+  @Mock
+  protected ThreadLocal<List<LogEntry>> mockedAsm;
+
+  /**
+   * Init mocked objects, particularly the authentication as the given county.
+   * @param countyName The name of the county - must match whatever is loaded into the test database.
+   * @param countyID   The ID of the county - must match whatever is loaded into the test database.
+   */
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/TestClassWithAuth.java:73]: Can we make countyName and countyID final?

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkResponseStub.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package au.org.democracydevelopers.corla.util;
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkResponseStub.java:23]: Class missing top level copyright notices etc.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package au.org.democracydevelopers.corla.util;
+
+import spark.Request;
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java:24]: Class missing top level copyright notices etc.

**Review by michelleblom — APPROVED** [review]

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 12]
```diff
@@ -98,8 +96,8 @@ public IRVBallotInterpretation() {
    * @param orderedChoices the way colorado-rla interpreted the raw choices, as an order list of names.
    */
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:96]: cvrNumber can be final (I think I made this comment on another PR).

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 13]
```diff
@@ -98,8 +96,8 @@ public IRVBallotInterpretation() {
...
   public IRVBallotInterpretation(final Contest contest, final CastVoteRecord.RecordType recordType,
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:97]: cvrNumber can be final

**Review by michelleblom — APPROVED** [review]

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java (hunk 7 above)
Comment (vteague) [reply server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/util/SparkRequestStub.java:35]: I have added some more in another branch, so I'm leaving this for now but will have another look when I merge the other PR.

**Pull request merged**